plugins/ltac2/g_ltac2.mlg

```diff
@@ -200,15 +212,23 @@ GRAMMAR EXTEND Gram
         { CAst.make ~loc @@ CTacCst (AbsKn (Tuple 0)) }
       | "("; ")" ->
         { CAst.make ~loc @@ CTacCst (AbsKn (Tuple 0)) }
-      | "["; a = LIST0 ltac2_expr LEVEL "5" SEP ";"; "]" ->
-        { Tac2quote.of_list ~loc (fun x -> x) a }
+      | a = array_literal -> { a }
+      | a = list_literal -> { a }
       | "{"; test_ident_with_or_lpar_or_rbrac; e = ltac2_expr LEVEL "0"; "with"; a = tac2rec_fieldexprs; "}" ->
         { CAst.make ~loc @@ CTacRec (Some e, a) }
       | "{"; a = tac2rec_fieldexprs; "}" ->
         { CAst.make ~loc @@ CTacRec (None, a) }
       | a = tactic_atom -> { a } ]
     ]
   ;
+  array_literal:
+  [ [ test_array_opening; "["; "|"; a = LIST0 ltac2_expr LEVEL "5" SEP ";"; test_array_closing; "|"; "]" ->
+      { Tac2quote.array_of_list (Tac2quote.of_list ~loc (fun x -> x) a) } ] ]
+  ;
+  list_literal:
+  [ [ "["; a = LIST0 ltac2_expr LEVEL "5" SEP ";"; "]" ->
+       { Tac2quote.of_list ~loc (fun x -> x) a } ] ]
+  ;
```

Comment:
How about inlining thes two productions in `ltac2_expr` instead of creating new nonterminals?  You could add short comments describing what they do.

Reply:
That's what I tried first, but it doesn't work: No matter in which order I add the two productions, it behaves as if there was no production to parse arrays. My suspicion is that when the rules get inserted, there is some code trying to be smart and to factorize the two rules that both start with `[`, but it's not quite smart enough. When I create new nonterminals, this factorization code does not run, and it works.

Reply:
The parser is not state of the art :-(.  There is special case code to handle a series of IDENTs at the beginning of a production so that multiword tactic and command names match correctly.  Other than that, once the first symbol in a production is recognized, the parser won't consider other productions for the nonterminal (e.g. `ltac2_expr` in this case).  Productions are processed sequentially.  It just might be that the mlg productions are added in reverse order: `list_literal` production may be processed first and would match on the initial `[`.  It _might_ work if you reverse the order - perhaps worth trying just for the sake of understanding.  But what you have is fine.

Reply:
I tried both inlined orders, but none worked.